src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Usage/DetectPLINQNopsAnalyzer.cs

```diff
@@ -0,0 +1,115 @@
+                                                                             DiagnosticCategory.Usage,
+                                                                             RuleLevel.IdeSuggestion,
+                                                                             description: s_localizableDescription,
+                                                                             isPortedFxCopRule: false,
+                                                                             isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(DefaultRule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(ctx =>
+            {
+                if (!ctx.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemLinqParallelEnumerable, out var parallelEnumerable))
+                {
+                    return;
+                }
+
+                if (!ctx.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemLinqEnumerable, out var linqEnumerable))
+                {
+                    return;
+                }
+
+                var asParallelSymbols = parallelEnumerable.GetMembers("AsParallel").ToImmutableHashSet();
+                var collectionSymbols = parallelEnumerable.GetMembers("ToArray")
+                .Concat(parallelEnumerable.GetMembers("ToList"))
+                .Concat(parallelEnumerable.GetMembers("ToDictionary"))
+                .Concat(linqEnumerable.GetMembers("ToHashSet"))
+                .ToImmutableHashSet();
+
+                ctx.RegisterOperationAction(x => AnalyzeOperation(x, asParallelSymbols, collectionSymbols), OperationKind.Invocation);
+            });
+        }
+
+        public static bool ParentIsForEachStatement(IInvocationOperation operation) => operation.Parent is IForEachLoopOperation || operation.Parent?.Parent is IForEachLoopOperation;
+
+        public static bool TryGetParentIsToCollection(IInvocationOperation operation, ImmutableHashSet<ISymbol> collectionSymbols, out IInvocationOperation parentInvocation)
+        {
+            parentInvocation = operation;
+            if (operation.Parent?.Parent is not IInvocationOperation invocation)
+            {
+                if (operation.Parent?.Parent?.Parent is IInvocationOperation nestedInvocation) // Happens on ToHashSet calls
+                {
+                    parentInvocation = nestedInvocation;
+                    return true;
+                }
+
+                return false;
+            }
+
+            if (collectionSymbols.Contains(invocation.TargetMethod.OriginalDefinition) || collectionSymbols.Contains(invocation.TargetMethod.OriginalDefinition.ReducedFrom))
```

Comment:
Don't know if there is any importance but I have seen that other analyzers usually check for `OriginalDefinition` after the `ReducedFrom`.

Reply:
OriginalDefinition worked for me in my C# tests, the second test was to make vb work.

Reply:
I think it is better to write this as `var targetMethod = (invocation.TargetMethod.ReducedFrom ?? invocation.TargetMethod).OriginalDefinition` and then operate on `targetMethod`.